### DISCUSSION TITLE
AutoScaling - Quarkus, KNative

### DISCUSSION THREAD

**Discussion by srinivasks22** [apache/camel-karavan#1002] [Q&A] [answered]
Using Karavan, Openshift, Camel Jang with below properties and trying to get to the default behaviour which is if the requests are idle, the Openshift would scale down the container to Zero. But as such it is not happenning. I tried explicitly setting the value quarkus.knative.scale-to-zero-enabled=true but still no luck? Any guidance would be helpful.

camel.health.enabled=true
camel.health.exposure-level=full
camel.karavan.project-id=karavan-serving
camel.karavan.project-name=karavan-serving
camel.karavan.project-description=karavan-serving
camel.karavan.target=openshift
camel.jbang.gav=org.kafka:karavan-serving:1
camel.jbang.quarkusVersion=3.5.2
camel.jbang.runtime=quarkus
camel.jbang.exportDir=.export
camel.jbang.dependencies=camel:microprofile-health,mvn:io.quarkus:quarkus-openshift,mvn:io.quarkus:quarkus-container-image-jib,mvn:io.quarkus:quarkus-resteasy-reactive,mvn:io.quarkus:quarkus-kubernetes
# image build properties
quarkus.container-image.build=true
quarkus.container-image.group=${NAMESPACE}
quarkus.container-image.name=interop-poc/karavan-serving
quarkus.container-image.tag=${DATE}
quarkus.container-image.builder=openshift
quarkus.container-image.registry=image-registry.openshift-image-registry.svc:5000
quarkus.openshift.resources.requests.memory=128Mi
quarkus.openshift.resources.requests.cpu=250m
quarkus.openshift.resources.limits.memory=256Mi
quarkus.openshift.resources.limits.cpu=1000m
# deployment properties
quarkus.kubernetes.deploy=true
quarkus.kubernetes-client.trust-certs=true
#quarkus.kubernetes.deployment-target=openshift
quarkus.kubernetes.deployment-target=knative
quarkus.kubernetes-client.namespace=${NAMESPACE}
quarkus.kubernetes.namespace=${NAMESPACE}
quarkus.openshift.labels."app.openshift.io/runtime"=camel
quarkus.openshift.deployment-kind=Deployment
quarkus.openshift.add-version-to-label-selectors=false
quarkus.openshift.route.expose=true
quarkus.openshift.part-of=karavan-serving
quarkus.openshift.replicas=1
quarkus.openshift.base-jvm-image=registry.access.redhat.com/ubi9/openjdk-17:1.17-1

##############
#quarkus.openshift.route.target-port=
quarkus.knative.min-scale=0
quarkus.knative.max-scale=5
quarkus.knative.scale-to-zero-enabled=true
quarkus.knative.revision-auto-scaling.auto-scaler-class=kpa
#quarkus.knative.revision-auto-scaling.metric=cpu

**Comment by mgubaidullin** [comment]
I am not an expert on Knative but I believe that it should be Openshift Serverless installed on OpenShift for that

**Comment by srinivasks22** [comment]
In my case the Red Hat OpenShift Serverless is already installed. One more thing to worth mention, with the above set of properties, I see a configMap that is getting created with empty keys/values.

not sure if all the quarkus.knative.* properties can be leveraged as part of the Karavan's project config? I was assuming that should be the case. But just wanted to mention atleast in my scenario, the below configs doesn't really seem to matter. Do I've to include any thing else in the camel.jbang.dependencies property?

quarkus.knative.min-scale=0
quarkus.knative.max-scale=5
quarkus.knative.scale-to-zero-enabled=true
quarkus.knative.revision-auto-scaling.auto-scaler-class=kpa

**Comment by davsclaus** [comment]
You need to export as a quarkus based project, karavan exports as camel-main only. And you need to adjust the build script that does the camel-jbang export

**Reply by srinivasks22** [reply]
I used VSStudio and while creating the project i selected quarkus

<img width="264" alt="Screen Shot 2023-11-27 at 10 02 47 AM" src="https://github.com/apache/camel-karavan/assets/2183549/2244646a-0ba7-4eb4-b95e-5ad9da99072d">

Also, is there a command to export the project as quarkus based? The one that I'm using right now is as follows : 
jbang --java 17 "-Dcamel.jbang.version=4.2.0" camel@apache/camel export --fresh  --directory=<project_location>/.export && mvn clean package -f .export

**Reply by srinivasks22** [reply]
@davsclaus - any thoughts?

**Reply by davsclaus** [reply]
The doc for exporting to quarkus is here
https://camel.apache.org/manual/camel-jbang.html#_exporting_to_camel_quarkus

**Reply by davsclaus** [reply]
Also see the quarkus docs for knative
https://quarkus.io/guides/deploying-to-openshift#knative-openshift-serverless

**Reply by srinivasks22** [reply]
As part of the application.properties in Karavan project, there is this below properties that I've already used. I don't think we have to explicitly specify the runtime version again. However with the below configs in place, the Knative stuff is not working as expected.

camel.jbang.quarkusVersion=3.5.2
**camel.jbang.runtime**=quarkus
camel.jbang.exportDir=.export

**Reply by davsclaus** [reply, marked as answer]
Well karavan does not support knative directly OOTB - You can export the project to quarkus, and try locally to see what it takes to make that work with your knative platform.

**Reply by srinivasks22** [reply]
Hmm...one thing I want to highlight. In the above configurations there is one property **quarkus.kubernetes.deployment-target=knative**. When I build the application (which comprises of just one camel rest component and logs it out ) and export it to Openshift, it is indeed deployed as serverless meaning I can see it. You can refer the screen shot below. This tells me that the packaging has been deployed as serverless for sure. However, I was trying to test if it is truly serverless by checking the default behavior where in, if the traffic is not active for more than 30 secs, the pod should scale down to zero which was not happening. 

<img width="664" alt="Screen Shot 2023-11-30 at 7 41 57 AM" src="https://github.com/apache/camel-karavan/assets/2183549/f4976080-103d-41b8-a327-2fd300e1cfb7">




On a separate stream, I've build KNative apps using Quarkus and deployed them successfully which were scaling down to zero. I want to check if Karavan packaging can also get me to the same behaviour and if so it would be a great feature.

**Reply by davsclaus** [reply]
Scale to zero depends on what camel components you use as the source(s).

**Reply by srinivasks22** [reply]
I'm using camel **rest**. I assume it can be scaled down to zero?
